lib/routes/syosetu/search.ts

```diff
@@ -0,0 +1,288 @@
+    searchParams.notword = setIfExists(params.notword);
+
+    searchParams.title = setIfExists(params.title);
+    searchParams.ex = setIfExists(params.ex);
+    searchParams.keyword = setIfExists(params.keyword);
+    searchParams.wname = setIfExists(params.wname);
+
+    searchParams.sasie = setIfExists(params.sasie);
+    searchParams.iszankoku = setIfExists(params.iszankoku);
+    searchParams.isbl = setIfExists(params.isbl);
+    searchParams.isgl = setIfExists(params.isgl);
+    searchParams.istensei = setIfExists(params.istensei);
+    searchParams.istenni = setIfExists(params.istenni);
+
+    searchParams.stop = setIfExists(params.stop);
+    searchParams.notzankoku = setIfExists(params.notzankoku);
+    searchParams.notbl = setIfExists(params.notbl);
+    searchParams.notgl = setIfExists(params.notgl);
+    searchParams.nottensei = setIfExists(params.nottensei);
+    searchParams.nottenni = setIfExists(params.nottenni);
+
+    searchParams.minlen = setIfExists(params.minlen);
+    searchParams.maxlen = setIfExists(params.maxlen);
+
+    searchParams.type = setIfExists(params.type as NovelTypeParam);
+    searchParams.order = setIfExists(params.order as Order);
+    searchParams.genre = setIfExists(params.genre as Join<Genre> | Genre);
+    searchParams.nocgenre = setIfExists(params.nocgenre as Join<R18Site> | R18Site);
+
+    if (params.mintime || params.maxtime) {
+        searchParams.time = `${params.mintime || ''}-${params.maxtime || ''}`;
+    }
+
+    return searchParams;
+}
+
+enum SyosetuSub {
+    NORMAL = 'yomou',
+    NOCTURNE = 'noc',
+    MOONLIGHT = 'mnlt',
+    MIDNIGHT = 'mid',
+}
+
+const isGeneral = (sub: string): boolean => sub === SyosetuSub.NORMAL;
+
+function createNovelSearchBuilder(sub: string, searchParams: SearchParams) {
+    if (isGeneral(sub)) {
+        return new SearchBuilder(searchParams, new NarouNovelFetch());
+    }
+
+    const r18Params = { ...searchParams };
+
+    switch (sub) {
+        case SyosetuSub.NOCTURNE:
+            r18Params.nocgenre = R18Site.Nocturne;
+            break;
+        case SyosetuSub.MOONLIGHT:
+            // If either 女性向け/BL is chosen, nocgenre will be in query string
+            // If no specific genre selected, include both
+            if (!r18Params.nocgenre) {
+                r18Params.nocgenre = [R18Site.MoonLight, R18Site.MoonLightBL].join('-') as Join<R18Site>;
+            }
+            break;
+        case SyosetuSub.MIDNIGHT:
+            r18Params.nocgenre = R18Site.Midnight;
+            break;
+        default:
+            throw new InvalidParameterError('Invalid Syosetu subsite.\nValid subsites are: yomou, noc, mnlt, mid');
+    }
+
+    return new SearchBuilderR18(r18Params, new NarouNovelFetch());
+}
+
+async function handler(ctx: Context): Promise<Data> {
+    const { sub, query } = ctx.req.param();
+    const searchUrl = `https://${sub}.syosetu.com/search/search/search.php?${query}`;
+
+    return (await cache.tryGet(searchUrl, async () => {
```

Comment:
Please make use of the [common parameter limit](https://docs.rsshub.app/guide/parameters#limit-entries) `ctx.req.query('limit')` instead.